### ISSUE TITLE
Determine what version of CVE database was used in a scan

### ISSUE THREAD

**Issue by CameronLonsdale** [anchore/anchore-engine#43]
I couldn't see this feature anywhere, I could be wrong, but if not I wanted to start a discussion of storing information about what the CVE database looked like at the time of a scan. This could be useful if you want to find out if a particular scan did or did not check for a CVE.

**Comment by acathrow** [comment]
The engine _scans_ the image once, which gathers all the data about the image - files, packages, metadata, etc. When the engine is asked to report on CVEs evaluate an image against a policy it users the latest vulnerability data available.
So there's really a _scan_ and a _report_ to consider the timing off.
If I add an image to the Anchore Engine now and immediately request a CVE report I'll get the CVEs in that image as of right now. If I request the CVE report tomorrow I'll get the latest list of matching CVEs.
The engine will update its list of CVEs every 6 hours (user configurable) and as new CVEs are added it will match them against images.
We have a subscription/notification system see: [here](https://anchore.freshdesk.com/support/solutions/articles/36000003890-working-with-subscriptions) which can be used to send out notifications based on key events. For example you can subscribe to receive a notification if new CVEs are added to a specific image.

Does that answer your question?

**Comment by CameronLonsdale** [comment]
Right, so I guess i'm thinking of the scenario where the user configures the CVE update to be once a week say. A CVE is added to a feed on Wednesday, you scan and evaluate an image against the policy on Thursday but you didn't in-fact check against the latest CVE because your local database wasn't updated. So when you get the policy evaluation results you could attach with it the last time the database was updated to be able to determine if you evaluated against the most up to date CVE list or not. You could get around this issue by just forcing a database sync before every time you evaluate the policy, then you can use the current time as the database 'version'. Though people do like their metadata, it's fairly minor but could be useful for retrospective analysis.

Let me know if what I tried to explain isn't clear enough.

**Comment by acathrow** [comment]
We want to make policy evaluation and CVE results as real-time as possible, I doubt folks would make the feed sync internal longer, usually they go shorter to pick things up ASAP. That said, even 6 hours is very good given the frequency of updates. One of the policy checks we have in our policy language actually checks for stale feed sync data to allow a user's policy checks to issue warnings or even fail based on stale data. We do have the ability to retrieve the last feed sync timestamp for a specific feed (debian:8, centos:7, etc) see https://anchore.freshdesk.com/support/solutions/articles/36000052809-feed-synchronization-  so perhaps you'd like to see that date in the output as extra metadata, would that work?

One thing we've discussed in the past, which I think aligns with your point, is if we can perform a policy check or CVE scan as per a certain date.
Right now we keep the image data indefinitely (you can prune it of course) so I could run a policy or CVE check on an image from 18 months ago, for example. This would create a report based on the new data we have (old image, new vuln data) which tells us how secure the image _was_ based on what we know now. But we've discussed adding the option to use the old data. It's academically interesting, but we didn't see enough interest yet to do anything with that.

**Comment by CameronLonsdale** [comment]
Yeah the last feed sync timestamp as extra metadata could work. It would be interesting to see if other users would find this feature useful as well. Your second point is definitely academically interesting, you could use it to find interesting info about image decay, average time to get popped if left unpatched and other things, would be neat!

**Comment by zhill** [comment]
CVE scans in Anchore are done when the image is finished analysis and added to the db and are updated in the same transaction as CVE updates moving forward to keep the db fully consistent for package-based CVEs.

It seems that what you really want to know is:

1. If I'm not seeing a CVE I expect to see in an image, is it present in the system and was it checked for image X.
2. When was CVE Y added/updated in the system.

Broadly, to address these issues I think we'll add:
1. API to query feed sync task state and history (when were feed syncs run, what was the result, and is one running now)
2. API to query the CVE/Feed db in the system to get CVEs by name, date, or by package/artifact

Those are already things I'm thinking about, but don't have a timeline for implementation yet. Would those things address your requirements?

**Comment by zengjunSpring** [comment]
生产环境没有网的情况下,怎么保证漏洞库是实时更新的???

**Comment by tuhaolam** [comment]
> 生产环境没有网的情况下,怎么保证漏洞库是实时更新的???

Please modify your problem using English. Thank You:)

**Comment by CameronLonsdale** [comment]
> 生产环境没有网的情况下,怎么保证漏洞库是实时更新的???

Using Google Translate: "How to ensure that the vulnerability library is updated in real time when there is no network in the production environment."

The way that I solved this problem is to have a separate instance of anchore running on a non prod environment which has internet access, trigger an update, snapshot the database and upload that to your production anchore. The downside is that this is not 'realtime' but it can be done as frequently as you like (once a day for example). 

This can also solve the problem of having the database 'versioned as you can hash the database, and know that it won't change in production as there is no internet access.
